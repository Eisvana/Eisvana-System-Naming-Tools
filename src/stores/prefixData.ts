import { useCoords2Glyphs } from '@/composables/coordGlyphConvert';
import { defineStore } from 'pinia';

interface State {
  glyphs: string;
  systemName: string;
  prefix: string;
}

export const usePrefixDataStore = defineStore('prefixData', {
  state: (): State => ({
    glyphs: '',
    systemName: '',
    prefix: '',
  }),

  getters: {
    glyphValues: (state) => useCoords2Glyphs(state.glyphs),
  },
});
