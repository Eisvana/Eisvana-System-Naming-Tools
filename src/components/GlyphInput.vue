<script setup lang="ts">
import { computed } from 'vue';
import { usePrefixDataStore } from '../stores/prefixData';
import { storeToRefs } from 'pinia';
import { maxCoordLength, maxGlyphLength, validInputRegex } from '../variables/constants';

const prefixDataStore = usePrefixDataStore();
const { glyphs, glyphValues } = storeToRefs(prefixDataStore);

defineProps<{
  error: boolean;
}>();

function addGlyph(e: Event) {
  if (!(e.target instanceof HTMLButtonElement)) return;
  glyphs.value ??= '';
  if (glyphs.value?.length < maxGlyphLength) {
    glyphs.value += e.target.value; // NoSonar 12 is maximum glyph length
  }
}

function deleteGlyph() {
  glyphs.value = glyphs.value.slice(0, -1);
}

function lintGlyphs() {
  glyphs.value = glyphs.value
    .toUpperCase()
    .split('')
    .filter((char) => validInputRegex.test(char))
    .join('');
}

const numberToGlyph = (n: number) => n.toString(16).toUpperCase(); // NoSonar this is dec to hex

const isCoords = computed(() => glyphValues.value.includes(':'));

// uses 19 if input is empty
const maxLength = computed(() => (isCoords.value || !glyphValues.value ? maxCoordLength : maxGlyphLength));
</script>

<template>
  <div>
    <label
      class="required"
      for="portalglyphsInput"
      >Glyphs or Coordinates:</label
    >
    <div class="glyph-input-wrapper">
      <input
        :aria-invalid="error || undefined"
        :maxlength="maxLength"
        class="glyphs-input"
        id="portalglyphsInput"
        type="text"
        v-model="glyphs"
        @input="lintGlyphs"
      />
      <button
        class="delete-button is-error"
        id="delButton"
        type="button"
        @click="deleteGlyph"
      >
        &larr; Delete
      </button>
    </div>
    <div class="portal-buttons grid">
      <button
        v-for="n in 16"
        class="button glyphs"
        type="button"
        :id="'glyphButton' + n"
        :value="numberToGlyph(n - 1)"
        @click="addGlyph"
      >
        {{ numberToGlyph(n - 1) }}
      </button>
    </div>
    <p
      v-show="glyphValues && !glyphValues.includes(':')"
      class="glyph-display-wrapper"
    >
      <output
        class="glyphs"
        id="glyphDisplay"
        >{{ glyphValues }}</output
      >
    </p>
  </div>
</template>

<style scoped lang="scss">
.glyphs {
  font-family: NMS-Glyphs-Mono;
  font-size: 3rem;
  word-break: break-word;
}

.glyph-display-wrapper {
  margin-bottom: 0 !important;
}

.portal-buttons {
  display: grid;
  grid-template-columns: repeat(8, auto);
  max-width: 800px;

  .button {
    line-height: 3rem;
    padding: 0;
    margin: 0;
  }
}

@media only screen and (max-width: 767px) {
  .portal-buttons {
    display: flex;
    flex-wrap: wrap;

    .button {
      width: 4rem;
    }
  }
}

.glyph-input-wrapper {
  display: flex;
  flex-wrap: wrap;
  max-width: 800px;

  .delete-button {
    margin-inline-start: 0.5rem;
    width: auto;
  }

  .glyphs-input {
    flex-grow: 1;
    width: auto;
  }
}

.is-error {
  background-color: indianred;
  border: 1px solid firebrick;
}
</style>
