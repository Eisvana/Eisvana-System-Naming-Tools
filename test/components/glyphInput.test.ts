import { VueWrapper, mount } from '@vue/test-utils';
import { usePrefixDataStore } from '@/stores/prefixData';
import GlyphInput from '@/components/GlyphInput.vue';
import { beforeEach, describe, expect, it } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import { storeToRefs } from 'pinia';

describe('GlyphInput', () => {
  let wrapper: VueWrapper;
  const pinia = createTestingPinia();

  beforeEach(() => {
    wrapper = mount(GlyphInput, {
      global: {
        plugins: [pinia],
      },
      props: {
        error: false,
      },
    });
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when component is in error', async () => {
    await wrapper.setProps({
      error: true,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('updates glyphs.value when a button is clicked', async () => {
    const store = usePrefixDataStore(pinia);
    const { glyphs } = storeToRefs(store);
    const button = wrapper.find('button.button.glyphs');
    await button.trigger('click');

    expect(glyphs.value).toBe('0');
  });

  it('does not update glyphs.value when a bad value is entered', async () => {
    const store = usePrefixDataStore(pinia);
    const { glyphs } = storeToRefs(store);
    const input = wrapper.find('input[type="text"]');
    await input.setValue('h');

    expect(glyphs.value).toBe('');
  });

  it('converts coords to glyphs', async () => {
    const store = usePrefixDataStore(pinia);
    const { glyphValues } = storeToRefs(store);
    const input = wrapper.find('input[type="text"]');
    await input.setValue('0FFE:0021:090F:005E');

    expect(glyphValues.value).toBe('005EA21107FF');
  });

  it('keeps glyphs intact', async () => {
    const store = usePrefixDataStore(pinia);
    const { glyphValues } = storeToRefs(store);
    const input = wrapper.find('input[type="text"]');
    await input.setValue('005EA21107FF');

    expect(glyphValues.value).toBe('005EA21107FF');
  });

  it('removes glyph', async () => {
    const store = usePrefixDataStore(pinia);
    const { glyphs } = storeToRefs(store);
    const input = wrapper.find('input[type="text"]');
    await input.setValue('0');

    const button = wrapper.find('button#delButton');
    await button.trigger('click');

    expect(glyphs.value).toBe('');
  });
});
