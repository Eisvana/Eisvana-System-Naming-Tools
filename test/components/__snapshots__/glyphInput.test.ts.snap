// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GlyphInput > should render correctly 1`] = `
"<div data-v-ed03a538=""><label data-v-ed03a538="" class="required" for="portalglyphsInput">Glyphs or Coordinates:</label>
  <div data-v-ed03a538="" class="glyph-input-wrapper"><input data-v-ed03a538="" maxlength="19" class="glyphs-input" id="portalglyphsInput" type="text"><button data-v-ed03a538="" class="delete-button is-error" id="delButton" type="button"> ← Delete </button></div>
  <div data-v-ed03a538="" class="portal-buttons grid"><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton1" value="0">0</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton2" value="1">1</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton3" value="2">2</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton4" value="3">3</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton5" value="4">4</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton6" value="5">5</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton7" value="6">6</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton8" value="7">7</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton9" value="8">8</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton10" value="9">9</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton11" value="A">A</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton12" value="B">B</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton13" value="C">C</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton14" value="D">D</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton15" value="E">E</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton16" value="F">F</button></div>
  <p data-v-ed03a538="" class="glyph-display-wrapper" style="display: none;"><output data-v-ed03a538="" class="glyphs" id="glyphDisplay"></output></p>
</div>"
`;

exports[`GlyphInput > should render correctly when component is in error 1`] = `
"<div data-v-ed03a538=""><label data-v-ed03a538="" class="required" for="portalglyphsInput">Glyphs or Coordinates:</label>
  <div data-v-ed03a538="" class="glyph-input-wrapper"><input data-v-ed03a538="" maxlength="19" class="glyphs-input" id="portalglyphsInput" type="text" aria-invalid="true"><button data-v-ed03a538="" class="delete-button is-error" id="delButton" type="button"> ← Delete </button></div>
  <div data-v-ed03a538="" class="portal-buttons grid"><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton1" value="0">0</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton2" value="1">1</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton3" value="2">2</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton4" value="3">3</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton5" value="4">4</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton6" value="5">5</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton7" value="6">6</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton8" value="7">7</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton9" value="8">8</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton10" value="9">9</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton11" value="A">A</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton12" value="B">B</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton13" value="C">C</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton14" value="D">D</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton15" value="E">E</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton16" value="F">F</button></div>
  <p data-v-ed03a538="" class="glyph-display-wrapper" style="display: none;"><output data-v-ed03a538="" class="glyphs" id="glyphDisplay"></output></p>
</div>"
`;
